### PR TITLE
Fix axis getting mixed up when split leaf

### PR DESCRIPTION
**Pull request by Rindbee**
`order[0]` should hold the **longest** axis (`max_axis_index()`), while `order[POINT::AXIS_COUNT - 1]` should hold the **shortest** axis (`min_axis_index()`).

### CODE REVIEW

**Comment by lawnjelly** [comment]
Bear in mind, there is a possibility that the comment is out of date (rather than the code wrong way). :smile: 

I did quite a bit of empirical testing at the time, so it would be nice to double check this is an improvement with some testing (I have vague memory some of the results were counter intuitive).

File: core/math/bvh_split.inc
Comment (lawnjelly) [inline core/math/bvh_split.inc:25]: Would expect optimizer to re-order this anyway (unless there was some conversion going on for the index), but the change is fine. :+1:

**Review by lawnjelly — APPROVED** [review]
I think on balance this is worth a go.
It does look like it may have been unintentional, and if the other way had been better, I should have changed the comment at the time.

So I think we should go for this and watch for any regressions. :+1: 
Also needs changing in 3.x.

(For reference for mergers, this is unlikely to cause bugs, as it just affects the optimality of the tree when reshaping, and thus performance.)

## Aside
As an aside it's nice to have another pair of eyes looking through this. It's quite specialized so doesn't get as much review as I would ideally like .. although @pouleyKetchoupp did spend quite a while working on it too. The area most in need of scrutiny if I remember right is the `_logic_balance` area as it depended on some of Randy Gaul's code.

I did in fact last week start writing a reference version as a kind of "unit test" so we could double check any physics bugs were physics side rather than BVH side. I haven't rolled this out but it didn't detect any bugs so far which is good.

I also think physics should have a reference brute force broad phase, both because of the possibility of bugs in BVH, but also because I have long suspected they are not taking proper account of tunneling. In BVH if you move AABB from A to B it will only consider collisions at A, and at B, rather than in between. Whereas for a move I suspect the physics should be passing a merged AABB of the current and previous AABB. BVH will not do this for the client, that is client responsibility.

**Comment by Rindbee** [comment]
> The area most in need of scrutiny if I remember right is the `_logic_balance` area as it depended on some of Randy Gaul's code.

`_logic_balance` seems to be fine, if we can ensure that the relevant nodes have been refitted. 

The areas that I think are problematic are those that use `node_remove_item()` to remove and re-insert.

After a leaf node with multiple items uses `node_remove_item()` to remove an item, the aabb of the leaf node is outdated. Before calling `refit_branch()`, the results obtained by using other `refit*` methods may also be dirty, that is, these `refit*` calls seem to have little meaning.

**Comment by akien-mga** [comment]
Thanks!

**Comment by lawnjelly** [comment]
> After a leaf node with multiple items uses node_remove_item() to remove an item, the aabb of the leaf node is outdated.

There is an optimization in `node_remove_item`, as recalculating the leaf AABB is expensive. If the removed node is totally within the leaf AABB (with some expansions), then it is not a "corner AABB", it does not push the extents, so there is no need to recalculate the leaf AABB.

**Comment by Rindbee** [comment]
Before, I was concerned that it might be missing something. Now it feels like some leaf nodes may be calculated multiple times.

- leaf nodes whose leaf marked dirty will be caught by `refit_branch`, and then recalculate these leaf nodes.  
- `_node_add_item` will recalculate the leaf node immediately;
- `split_leaf` will create two new leaf node (their leaves will be marked dirty, for recalculate?) ;

It seems that `clear` should mark `dirty` as `false`; only in some special cases, `node_remove_item` should mark `dirty` as `true`.

```diff
--- a/core/math/bvh_structs.inc
+++ b/core/math/bvh_structs.inc
@@ -83,7 +83,7 @@ public:
 
        void clear() {
                num_items = 0;
-               set_dirty(true);
+               set_dirty(false);
        }
```

# Aside

I'm confused by the results from [test_perf_broadphase](https://github.com/godotengine/godot-demo-projects/blob/master/2d/physics_tests/tests/performance/test_perf_broadphase.tscn). While it feels like some expected stages are shortening, others are increasing. It feels like the modification shouldn't have much effect, but it has a significant effect.

Especially the Remove Time. A few days ago, testing 600x600 on my computer would take around 43s, now it takes around 30s.

**Comment by lawnjelly** [comment]
> Before, I was concerned that it might be missing something. Now it feels like some leaf nodes may be calculated multiple times.

Calculated in multiple places? Yes they can be. But ideally we don't want to calculate them multiple times though, if possible. But this may occur in some situations.

> leaf nodes whose leaf marked dirty will be caught by refit_branch, and then recalculate these leaf nodes.

Yes.

> `_node_add_item` will recalculate the leaf node immediately

`_node_add_item` returns a bool whether to refit, and like `node_remove_item`, it only orders a refit if the item added is a "corner item".

If I remember right, the general idea is this:
* If a leaf AABB is expanded (by adding a border item), it should be calculated immediately.
* Removing a non-border item does not affect the AABB, no calculation
* Removing a border item shrinks the AABB, so it is marked as dirty, to be recalculated on the next pass

The AABB is thus a _conservative bound_.
In some cases expanding an AABB could be deferred, but most subsequent operations require the conservative AABB to be up to date.

> split_leaf will create two new leaf node (their leaves will be marked dirty, for recalculate?)

Yes.

> It seems that clear should mark dirty as false; only in some special cases, node_remove_item should mark dirty as true.

`clear` sets the number of items in a node to zero, therefore the currently stored AABB is out of date, thus the node is marked as dirty.

> I'm confused by the results from [test_perf_broadphase](https://github.com/godotengine/godot-demo-projects/blob/master/2d/physics_tests/tests/performance/test_perf_broadphase.tscn). While it feels like some expected stages are shortening, others are increasing. It feels like the modification shouldn't have much effect, but it has a significant effect.
> Especially the Remove Time. A few days ago, testing 600x600 on my computer would take around 43s, now it takes around 30s.

I'm not sure I understand what you are saying here, I'll try this test and see if I can work it out. You talk about a modification, which modification? Modifying `clear()`, or longest axis, or something else?

**Comment by Rindbee** [comment]
> > It seems that clear should mark dirty as false; only in some special cases, node_remove_item should mark dirty as true.
> 
> `clear` sets the number of items in a node to zero, therefore the currently stored AABB is out of date, thus the node is marked as dirty.

Currently, the item will be added immediately after `clear`, which will immediately update the aabb of the leaf node. So it seems unnecessary for `clear` to mark `dirty` as `true` (and in most cases it is required that `num_items > 0`, the leaf node as the root node is an exception).

> Do you mean "creating objects, adding objects, moving objects"? This is unclear.

Yes. The stages recorded in the output.

> What modification are you referencing?

That's the result before and after applying this PR. I haven't figured out why yet.

> What has changed in these few days?

I'm sure it should be before and after this PR is merged. I verified it by reverting the commit. 

Some other interesting tests, they all have a greater impact on the time of "Removing objects":

Test 1.

https://github.com/godotengine/godot/blob/4c3dc26367518e006f8555c12f5d2df0b8a28192/servers/physics_2d/godot_collision_object_2d.cpp#L231


Similarly, revert this PR, and then replace the above code with
```cpp
if (!pending_shape_update_list.in_list()) {
	GodotPhysicsServer2D::godot_singleton->pending_shape_update_list.add(&pending_shape_update_list);
}
```

Test 2.

In the script, [here](https://github.com/godotengine/godot-demo-projects/blob/65b34f81920752a382d14d544aa451de46b32a07/2d/physics_tests/tests/performance/test_perf_broadphase.gd#L153-L154), if removed in the reverse order, it takes less time.

**Comment by Rindbee** [comment]
> I'm not familiar with this test, it seems one of pouleyKetchoups. I'm trying it out but get nothing on screen (maybe it deliberately doesn't render). Perhaps you can describe what this test does and what aspect of it you find interesting and why?

I'm not particularly familiar with it either, I just started learning about bvh last week and I'm not sure if these tests are appropriate.
It seems to add objects without overlapping, change the object's position, and then remove the object. It seems that the location and size are not random enough.

**Comment by lawnjelly** [comment]
> Currently, the item will be added immediately after clear, which will immediately update the aabb of the leaf node. So it seems unnecessary for clear to mark dirty as true (and in most cases it is required that num_items > 0, the leaf node as the root node is an exception).

I don't understand, is marking it as dirty causing a problem, or was this suggesting an optimization, or am I missing a call to set this as non-dirty afterwards?

> That's the result before and after applying this PR. I haven't figured out why yet.

Especially in artificial benchmarks, changes in the tree structure could lead to more noticeable changes in performance. The test in question seems to create items in a regular grid as far as I can see. Now whether any change in timing is due to directly the BVH or due to more collisions being processed by the physics is a question which can be revealed by profiling. (This is part of the reason the physics should also have a switchable reference broad phase imo.)

>  if removed in the reverse order, it takes less time.

This is generally the case due to the scene tree processing. This is why "global timings" often end up being difficult to interpret. It is important not to conflate changes in timing caused by other parts of the engine, with parts that are trying to be examined. Using several timers for different parts of the code can help here, and of course profiling.

**Comment by Rindbee** [comment]
> I don't understand, is marking it as dirty causing a problem, or was this suggesting an optimization, or am I missing a call to set this as non-dirty afterwards?

There seems to be no big problem. I thought that setting `dirty` to `false` would slightly reduce the time, but the result was not much improved. The number of leaf nodes requiring calls to `refit_upward` is indeed reduced. I feel that `refit_upward` is not particularly expensive. Maybe the generated aabb is not random enough?

**Comment by Rindbee** [comment]
| Before | After |
| :-------: | :------: |
| `set_dirty(true);` | `set_dirty(false);` |
|  |  |

**Comment by lawnjelly** [comment]
I've spent a little time this morning testing the splitting and it seems mostly ok. It is possible that we can set the child leaves to non-dirty after the split, as their AABB seems to be up to date at the end of the splitting routine.

This could be causing a single extra un-needed refit at a later stage, but it's probably not registered as a performance problem so far because:
1) Splitting is rare
2) Refits (without recalculating leaf AABBs which are expensive) are relatively cheap

**Comment by Rindbee** [comment]
This seems to be a bug.
https://github.com/godotengine/godot/blob/4c3dc26367518e006f8555c12f5d2df0b8a28192/core/math/bvh_refit.inc#L137

`p_node_id` should be `rp.node_id`.

**Comment by lawnjelly** [comment]
> This seems to be a bug.
> 
> https://github.com/godotengine/godot/blob/4c3dc26367518e006f8555c12f5d2df0b8a28192/core/math/bvh_refit.inc#L137
> 
> `p_node_id` might be `rp.node_id`.

Ah yes well spotted, that doesn't look right.

**Comment by Rindbee** [comment]
The number of splits is related to the number of items (`_refs.used_size()`).

It may be a little slower when adding items because the items will be moved (removed and reinserted),

The results in https://github.com/godotengine/godot/pull/82436#issuecomment-1738727945 are uncommon because aabb in actual cases may be more random.

**Pull request merged**
